File: src/homepod-base/flow.ts
import { Shortcuts } from '@basmilius/homey-common';
import type Homey from 'homey';
import type { AppleApp, HomePodBaseDevice, HomePodBaseDriver } from '../types';
import type { SoundBoardSound } from '../utils';

export default class HomePodFlow extends Shortcuts<AppleApp> {
    register(): void {
        this.#registerPlaySoundboard();
        this.#registerPlayUrl();
        this.#registerPlayUrlAtVolume();
        this.#registerSetPosition();
        this.#registerSkipBackward();
        this.#registerSkipForward();
    }

    async triggerArtworkUrlUpdated(device: HomePodBaseDevice<HomePodBaseDriver>, localUrl: string, cloudUrl: string): Promise<void> {
        try {
            const triggerCard = this.flow.getDeviceTriggerCard('homepod_artwork_url_updated');

            await triggerCard.trigger(device, {
                localUrl,
                cloudUrl
            });
        } catch (err) {
            this.log(device.name, 'Failed to trigger artwork url updated card.', err);
        }
    }

    #registerPlaySoundboard(): void {
        const card = this.homey.flow.getActionCard('homepod_play_soundboard');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly sound: SoundBoardSound;
            readonly volume: number;
        };

        card.registerRunListener(async ({device, sound, volume}: RunArguments) => {
            await device.playSoundboard(sound, volume);
        });

        card.registerArgumentAutocompleteListener('sound', async (query: string): Promise<Homey.FlowCard.ArgumentAutocompleteResults> => {
            const sounds = await this.app.soundBoard.getSounds();
            const lowerQuery = query.trim().toLowerCase();

            return sounds
                .filter(sound => lowerQuery.length === 0 || sound.name.toLowerCase().includes(lowerQuery))
                .sort((a, b) => a.name.localeCompare(b.name));
        });
    }

    #registerPlayUrl(): void {
        const playUrl = this.homey.flow.getActionCard('homepod_play_url');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly url: string;
        };

        playUrl.registerRunListener(async ({device, url}: RunArguments) => {
            await device.playUrl(url);
        });
    }

    #registerPlayUrlAtVolume(): void {
        const playUrl = this.homey.flow.getActionCard('homepod_play_url_at_volume');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly url: string;
            readonly volume: number;
        };

        playUrl.registerRunListener(async ({device, url, volume}: RunArguments) => {
            await device.playUrl(url, volume);
        });
    }

    #registerSetPosition(): void {
        const card = this.flow.getActionCard('homepod_set_position');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly position: number;
        };

        card.registerRunListener(async ({device, position}: RunArguments) => {
            await device.sdk.playback.seekTo(Math.floor(position));
        });
    }

    #registerSkipBackward(): void {
        const card = this.flow.getActionCard('homepod_skip_backward');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly seconds: number;
        };

        card.registerRunListener(async ({device, seconds}: RunArguments) => {
            await device.sdk.playback.skipBackward(Math.floor(seconds));
        });
    }

    #registerSkipForward(): void {
        const card = this.flow.getActionCard('homepod_skip_forward');

        type RunArguments = {
            readonly device: HomePodBaseDevice<HomePodBaseDriver>;
            readonly seconds: number;
        };

        card.registerRunListener(async ({device, seconds}: RunArguments) => {
            await device.sdk.playback.skipForward(Math.floor(seconds));
        });
    }
}

File: src/homepod_base/flow.py
from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any

import pyatv.interface as pyatv_interface

if TYPE_CHECKING:
    from ..app import AppleApp
    from .device import HomePodBaseDevice


def _require_atv(args: dict[str, Any], label: str) -> tuple[HomePodBaseDevice, pyatv_interface.AppleTV]:
    """Extract device and active pyatv connection from flow args, or raise."""
    device: HomePodBaseDevice = args['device']
    atv = device.atv
    if atv is None:
        raise RuntimeError(f'{label} "{device.get_name()}" is not connected')
    return device, atv


class HomePodFlow:
    """Registers and manages HomePod flow cards."""

    def __init__(self, app: AppleApp) -> None:
        self._app = app

    def register(self) -> None:
        self._register_play_url()
        self._register_play_url_at_volume()
        self._register_set_position()
        self._register_skip_forward()
        self._register_skip_backward()

    async def trigger_artwork_url_updated(
        self,
        device: HomePodBaseDevice,
        local_url: str,
        cloud_url: str,
    ) -> None:
        try:
            card = self._app.homey.flow.get_device_trigger_card(
                'homepod_artwork_url_updated'
            )
            await card.trigger(device, {'localUrl': local_url, 'cloudUrl': cloud_url})
        except asyncio.CancelledError:
            raise
        except Exception as err:
            self._app.log(device.get_name(), 'Failed to trigger artwork url updated card.', err)

    # ------------------------------------------------------------------
    # Action card registrations
    # ------------------------------------------------------------------

    def _register_play_url(self) -> None:
        card = self._app.homey.flow.get_action_card('homepod_play_url')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            device: HomePodBaseDevice = args['device']
            url: str = args['url']
            await device.play_url(url)

        card.register_run_listener(run)

    def _register_play_url_at_volume(self) -> None:
        card = self._app.homey.flow.get_action_card('homepod_play_url_at_volume')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            device: HomePodBaseDevice = args['device']
            url: str = args['url']
            volume: float = args['volume']
            await device.play_url(url, volume)

        card.register_run_listener(run)

    def _register_set_position(self) -> None:
        card = self._app.homey.flow.get_action_card('homepod_set_position')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'HomePod')
            await atv.remote_control.set_position(int(float(args['position'])))

        card.register_run_listener(run)

    def _register_skip_forward(self) -> None:
        card = self._app.homey.flow.get_action_card('homepod_skip_forward')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'HomePod')
            await atv.remote_control.skip_forward(int(float(args['seconds'])))

        card.register_run_listener(run)

    def _register_skip_backward(self) -> None:
        card = self._app.homey.flow.get_action_card('homepod_skip_backward')

        async def run(args: dict[str, Any], **kwargs: Any) -> None:
            _, atv = _require_atv(args, 'HomePod')
            await atv.remote_control.skip_backward(int(float(args['seconds'])))

        card.register_run_listener(run)
